libs/community/langchain_community/chat_models/outlines.py

```diff
@@ -0,0 +1,564 @@
+            ?term: factor (("*" | "/") factor)*
+            ?factor: NUMBER | "-" factor | "(" expression ")"
+            %import common.NUMBER
+        '''
+    
+    Note: Grammar-based generation is currently experimental and may have performance
+    limitations. It uses greedy generation to mitigate these issues.
+    
+    For more details and examples, see:
+    https://dottxt-ai.github.io/outlines/reference/generation/cfg/
+    """
+
+    custom_generator: Optional[Any] = None
+    """Set your own outlines generator object to override the default behavior."""
+
+    model_kwargs: Dict[str, Any] = Field(default_factory=dict)
+    """Additional parameters to pass to the underlying model.
+    
+    Example:
+        model_kwargs = {"temperature": 0.8, "seed": 42}
+    """
+
+    @model_validator(mode="after")
+    def validate_environment(self) -> "ChatOutlines":
+        """Validate that outlines is installed and create a model instance."""
+        try:
+            import outlines.models as models
+        except ImportError:
+            raise ImportError(
+                "Could not import the Outlines library. "
+                "Please install it with `pip install outlines`."
+            )
+
+        model: str = self.model
+        backend: str = self.backend
+
+        num_constraints = sum(
+            [
+                bool(self.regex),
+                bool(self.type_constraints),
+                bool(self.json_schema),
+                bool(self.grammar),
+            ]
+        )
+        if num_constraints > 1:
+            raise ValueError(
+                "Either none or exactly one of regex, type_constraints, "
+                "json_schema, or grammar can be provided."
+            )
+
+        def check_packages_installed(
+            packages: List[Union[str, Tuple[str, str]]],
+        ) -> None:
+            missing_packages = [
+                pkg if isinstance(pkg, str) else pkg[0]
+                for pkg in packages
+                if importlib.util.find_spec(pkg[1] if isinstance(pkg, tuple) else pkg)
+                is None
+            ]
+            if missing_packages:
+                raise ImportError(
+                    f"Missing packages: {', '.join(missing_packages)}. "
+                    "You can install them with:\n\n"
+                    f"    pip install {' '.join(missing_packages)}"
+                )
+
+        if backend == "llamacpp":
+            check_packages_installed([("llama-cpp-python", "llama_cpp")])
+            if ".gguf" in model:
+                creator, repo_name, file_name = model.split("/", 2)
+                repo_id = f"{creator}/{repo_name}"
+            else:
+                raise ValueError("GGUF file_name must be provided for llama.cpp.")
+            self.client = models.llamacpp(repo_id, file_name, **self.model_kwargs)
+        elif backend == "transformers":
+            check_packages_installed(["transformers", "torch", "datasets"])
```

Comment:
doesn't transformers install `torch` by default? (and maybe `datasets` too)

Reply:
https://github.com/huggingface/transformers/blob/b53e44e847230a39caecc320527776ed80a3ceff/setup.py#L279-L287
they are extras